### PR TITLE
Added to primitive iterables toArray() method which takes an array parameter

### PR DESCRIPTION
**Pull request by vmzakharov**
If the collection fits into the provided array it is used to store its elements and is returned from the method, otherwise a new array of the appropriate size is allocated and returned. If the iterable is empty, the target array is returned unchanged.

Signed-off-by: vmzakharov <zakharov.vladimir.m@gmail.com>

### CODE REVIEW

File: eclipse-collections-code-generator/src/main/resources/api/primitiveIterable.stg [hunk 1]
```diff
@@ -63,6 +63,13 @@ public interface <name>Iterable extends PrimitiveIterable
      */
     <type>[] toArray();
 
+    /**
+     * Converts the <name>Iterable to a primitive <type> array. If the collection fits into the provided array it is used
+     * to store its elements and is returned from the method, otherwise a new array of the appropriate size is allocated
+     * and returned. If the iterable is empty, the target array is returned unchanged.
+     */
+    <type>[] toArray(<type>[] target);
```
Comment (donraab) [inline eclipse-collections-code-generator/src/main/resources/api/primitiveIterable.stg:71]: If you implement this as a default method, then we can release in a minor version.

**Comment by donraab** [comment]
Suggested one change which will make this PR able to be merged for a minor release.

File: eclipse-collections-code-generator/src/main/resources/impl/bag/mutable/primitiveHashBag.stg [hunk 2]
```diff
@@ -745,6 +745,28 @@ public class <name>HashBag
...
+
+        <type>[] finalBypass = array;
+        this.forEachWithOccurrences((<type> each, int occurrences) ->
+        {
+            for (int i = 0; i \< occurrences; i++)
+            {
+                finalBypass[index[0]] = each;
+                index[0]++;
+            }
+        });
```
Comment (motlin) [inline eclipse-collections-code-generator/src/main/resources/impl/bag/mutable/primitiveHashBag.stg:766]: Nitpick: I don't see much advantage of using forEachWithOccurrences when the output is of size `size()`. Might as well delegate to forEachWithIndex or forEach?

File: eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveArrayList.stg [hunk 3]
```diff
@@ -92,6 +92,12 @@ final class Immutable<name>ArrayList
         return new Immutable<name>ArrayList(newArray);
     }
 
+    public static Immutable<name>ArrayList newListFrom(<name>List itemList)
+    {
+        <type>[] newArray = itemList.toArray(new <type>[itemList.size()]);
+        return new Immutable<name>ArrayList(newArray);
+    }
```
Comment (motlin) [inline eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveArrayList.stg:99]: Is this an unrelated change that snuck in?

Comment (donraab) [reply eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveArrayList.stg:99]: It's actually the reason for the change I believe.

Comment (donraab) [reply eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveArrayList.stg:99]: This could be clarified in the commit message or be a separate commit to split the use case from the change.

Comment (vmzakharov) [reply eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveArrayList.stg:99]: @motlin - a related change that snuck in. I will take it out from this PR.  @donraab is correct, there will be a separate change to take advantage of the toArray(target) in factory methods.
